### PR TITLE
reduce number of R calls in CharacterVectorOrderer

### PR DESCRIPTION
**Pull request by romainfrancois**
Addresses #3044.

### CODE REVIEW

File: src/api.cpp [hunk 1]
```diff
@@ -291,15 +291,15 @@ CharacterVectorOrderer::CharacterVectorOrderer(const CharacterVector& data) :
   LOG_VERBOSE << "Sorting " <<  n_uniques << " unique character elements";
 
   CharacterVector uniques(set.begin(), set.end());
-  CharacterVector s_uniques = Language("sort", uniques).fast_eval();
 
   // order the uniques with a callback to R
-  IntegerVector o = r_match(uniques, s_uniques);
+  IntegerVector o = Language("order", uniques, _["na.last"] = true).fast_eval();
```
Comment (krlmlr) [inline src/api.cpp]: `na.last` defaults to `TRUE`, but it's ok to be explicit here.

**Comment by krlmlr** [comment]
``` r
sort(c(1, 3, NA, 2))
#> [1] 1 2 3
```

Created on 2018-04-24 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

I think we need to tweak the code, which then also saves a comparison in the inner loop.

**Comment by krlmlr** [comment]
Only *dplyr.benchmark*, but I don't know which tests use the `CharacterVectorOrderer`. I think `arrange()` will hit this code path, maybe run on a column with values

```r
strsplit(paste(sample(letters, 1e7, replace = TRUE), collapse = ""), "z")[[1]]
```

**Comment by romainfrancois** [comment]
I probably should have started there. this is a bit disappointing: 

```r
> u <- unique(strsplit(paste(sample(letters, 1e6, replace = TRUE), collapse = ""), "z")[[1]])
> microbenchmark( match(u, sort(u)), order(u) )
Unit: milliseconds
              expr      min       lq     mean   median       uq      max neval
 match(u, sort(u)) 120.0047 122.0302 123.5795 123.2734 124.2714 133.3828   100
          order(u) 123.0409 124.4863 127.0454 125.8372 127.4428 143.2621   100
> 
> u <- unique(strsplit(paste(sample(letters, 1e7, replace = TRUE), collapse = ""), "z")[[1]])
> microbenchmark( match(u, sort(u)), order(u) )
Unit: seconds
              expr      min       lq     mean   median       uq      max neval
 match(u, sort(u)) 2.072618 2.159371 2.187151 2.172609 2.206047 2.379781   100
          order(u) 2.399566 2.418039 2.455926 2.427990 2.462395 2.726127   100
```

**Comment by romainfrancois** [comment]
I'll revisit later, but this feels very low priority

**Comment by krlmlr** [comment]
Pity. The other question is why `order()` seems slower than `sort()` + `match()`.

If we keep the old code, we should add a comment.

**Comment by krlmlr** [comment]
Maybe the timing has changed with R 3.5?

**Comment by romainfrancois** [comment]
I'm not on 3.5 yet

**Comment by krlmlr** [comment]
Edited the issue description to add a reference.

**Comment by romainfrancois** [comment]
Closing the pr now, but leaving the branch and the issue open.

**Pull request closed without merging**